File: javascript/day15_part1_2022.js
const fs = require('fs');

function abs(n) {
    return n < 0 ? -n : n;
}

function manhattan(p, q) {
    return abs(p.x - q.x) + abs(p.y - q.y);
}

function setOf(a) {
    const s = new Set();
    for (const c of a) {
        s.add(c);
    }
    return s;
}

function main() {
    const input = fs.readFileSync('input.txt', 'utf8').split('\n');
    const sensors = [];

    for (const line of input) {
        const parts = line.match(/Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/);
        const s = {
            pos: { x: parseInt(parts[1]), y: parseInt(parts[2]) },
            beacon: { x: parseInt(parts[3]), y: parseInt(parts[4]) }
        };
        s.dist = manhattan(s.pos, s.beacon);
        sensors.push(s);
    }

    console.log(impossible(sensors, 2000000));
}

function impossible(sensors, y) {
    const pts = setOf([]);
    for (const s of sensors) {
        const dist = s.dist - abs(s.pos.y - y);
        for (let x = 0; x <= dist; x++) {
            pts.add(s.pos.x + x);
            pts.add(s.pos.x - x);
        }
    }
    for (const s of sensors) {
        if (s.beacon.y === y) {
            pts.delete(s.beacon.x);
        }
    }
    return pts.size;
}

main();

File: julia/day15_part1_2022.jl

struct Sensor
    pos::Tuple{Int, Int}
    beacon::Tuple{Int, Int}
    dist::Int
end

function read_input(file_path)
    open(file_path, "r") do file
        return read(file, String)
    end
end

function manhattan(p, q)
    return abs(p[1] - q[1]) + abs(p[2] - q[2])
end

function impossible(sensors, y)
    pts = Set{Int}()
    for s in sensors
        dist = s.dist - abs(s.pos[2] - y)
        if dist >= 0
            for x in s.pos[1]-dist:s.pos[1]+dist
                push!(pts, x)
            end
        end
    end
    for s in sensors
        if s.beacon[2] == y
            delete!(pts, s.beacon[1])
        end
    end
    return length(pts)
end

function main()
    input = read_input("input.txt")
    sensors = Sensor[]
    for line in split(input, "\n")
        m = match(r"Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)", line)
        pos = (parse(Int, m.captures[1]), parse(Int, m.captures[2]))
        beacon = (parse(Int, m.captures[3]), parse(Int, m.captures[4]))
        dist = manhattan(pos, beacon)
        push!(sensors, Sensor(pos, beacon, dist))
    end
    println(impossible(sensors, 2000000))
end

main()
